File: backend/src/test/kotlin/com/aicoach/service/AthleteServiceTest.kt
package com.aicoach.service

import com.aicoach.domain.dto.CreateAthleteRequest
import com.aicoach.domain.dto.UpdateAthleteRequest
import com.aicoach.domain.entity.Athlete
import com.aicoach.domain.entity.FitnessLevel
import com.aicoach.domain.entity.GoalType
import com.aicoach.repository.AthleteRepository
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.springframework.web.server.ResponseStatusException
import java.util.Optional

class AthleteServiceTest : DescribeSpec({

    val athleteRepository = mockk<AthleteRepository>()
    val service = AthleteService(athleteRepository)

    fun makeAthlete(id: Long = 1L, name: String = "Alice") = Athlete(
        id = id,
        name = name,
        email = "alice@example.com",
        fitnessLevel = FitnessLevel.INTERMEDIATE,
        goalType = GoalType.TARGET_TIME,
        currentWeeklyKm = 60.0
    )

    describe("listAthletes") {
        it("returns mapped DTOs for all athletes") {
            val athletes = listOf(makeAthlete(1L, "Alice"), makeAthlete(2L, "Bob"))
            every { athleteRepository.findAll() } returns athletes

            val result = service.listAthletes()

            result.size shouldBe 2
            result[0].name shouldBe "Alice"
            result[1].name shouldBe "Bob"
        }
    }

    describe("getAthlete") {
        it("returns DTO when athlete is found") {
            val athlete = makeAthlete()
            every { athleteRepository.findById(1L) } returns Optional.of(athlete)

            val result = service.getAthlete(1L)

            result.id shouldBe 1L
            result.name shouldBe "Alice"
            result.fitnessLevel shouldBe FitnessLevel.INTERMEDIATE
        }

        it("throws 404 when athlete not found") {
            every { athleteRepository.findById(99L) } returns Optional.empty()

            val ex = shouldThrow<ResponseStatusException> { service.getAthlete(99L) }
            ex.statusCode.value() shouldBe 404
        }
    }

    describe("createAthlete") {
        it("saves entity and returns DTO with all fields set") {
            val request = CreateAthleteRequest(
                name = "Bob",
                email = "bob@example.com",
                fitnessLevel = FitnessLevel.ADVANCED,
                goalType = GoalType.PODIUM,
                currentWeeklyKm = 80.0,
                trainingDaysPerWeek = 5
            )
            val saved = Athlete(
                id = 2L,
                name = request.name,
                email = request.email,
                fitnessLevel = request.fitnessLevel,
                goalType = request.goalType,
                currentWeeklyKm = request.currentWeeklyKm,
                trainingDaysPerWeek = request.trainingDaysPerWeek
            )
            every { athleteRepository.save(any()) } returns saved

            val result = service.createAthlete(request)

            result.id shouldBe 2L
            result.name shouldBe "Bob"
            result.email shouldBe "bob@example.com"
            result.fitnessLevel shouldBe FitnessLevel.ADVANCED
            result.goalType shouldBe GoalType.PODIUM
            verify { athleteRepository.save(any()) }
        }
    }

    describe("updateAthlete") {
        it("applies partial update — only non-null fields") {
            val athlete = makeAthlete()
            every { athleteRepository.findById(1L) } returns Optional.of(athlete)
            every { athleteRepository.save(athlete) } returns athlete

            val request = UpdateAthleteRequest(name = "Alice Updated", currentWeeklyKm = 70.0)
            service.updateAthlete(1L, request)

            athlete.name shouldBe "Alice Updated"
            athlete.currentWeeklyKm shouldBe 70.0
            athlete.email shouldBe "alice@example.com" // unchanged
        }

        it("throws 404 when athlete not found") {
            every { athleteRepository.findById(99L) } returns Optional.empty()

            val ex = shouldThrow<ResponseStatusException> {
                service.updateAthlete(99L, UpdateAthleteRequest(name = "X"))
            }
            ex.statusCode.value() shouldBe 404
        }
    }

    describe("addCoachNote") {
        it("appends note with newline when existing notes present") {
            val athlete = makeAthlete().also { it.coachNotes = "Existing note" }
            every { athleteRepository.findById(1L) } returns Optional.of(athlete)
            every { athleteRepository.save(athlete) } returns athlete

            service.addCoachNote(1L, "New note")

            athlete.coachNotes shouldBe "Existing note\nNew note"
        }

        it("sets note directly when coach notes are blank") {
            val athlete = makeAthlete().also { it.coachNotes = null }
            every { athleteRepository.findById(1L) } returns Optional.of(athlete)
            every { athleteRepository.save(athlete) } returns athlete

            service.addCoachNote(1L, "First note")

            athlete.coachNotes shouldBe "First note"
        }
    }

    describe("linkStravaAthlete") {
        it("sets stravaAthleteId and stravaEnabled = true") {
            val athlete = makeAthlete()
            every { athleteRepository.findById(1L) } returns Optional.of(athlete)
            every { athleteRepository.save(athlete) } returns athlete

            service.linkStravaAthlete(1L, 42L)

            athlete.stravaAthleteId shouldBe 42L
            athlete.stravaEnabled shouldBe true
        }

        it("throws 404 when athlete not found") {
            every { athleteRepository.findById(99L) } returns Optional.empty()

            val ex = shouldThrow<ResponseStatusException> {
                service.linkStravaAthlete(99L, 42L)
            }
            ex.statusCode.value() shouldBe 404
        }
    }
})

File: server/src/services/athleteService.test.ts
import { describe, it, expect, beforeAll, afterEach } from 'vitest'
import { createTestDb, resetDb } from '../test/dbHelper.js'
import {
  listAthletes,
  getAthlete,
  createAthlete,
  updateAthlete,
  addCoachNote,
  deleteAthlete,
} from './athleteService.js'

beforeAll(async () => {
  await createTestDb()
})

afterEach(() => {
  resetDb()
})

describe('createAthlete', () => {
  it('creates athlete with name only; all optional fields null', () => {
    const athlete = createAthlete({ name: 'Alice' })
    expect(athlete.name).toBe('Alice')
    expect(athlete.email).toBeNull()
    expect(athlete.fitnessLevel).toBeNull()
    expect(athlete.experienceYears).toBeNull()
    expect(athlete.currentWeeklyKm).toBeNull()
    expect(athlete.trailAccess).toBe(false)
    expect(athlete.id).toBeGreaterThan(0)
  })

  it('trailAccess=true stored and returned as true', () => {
    const athlete = createAthlete({ name: 'Bob', trailAccess: true })
    expect(athlete.trailAccess).toBe(true)
  })

  it('trailAccess=false stored and returned as false', () => {
    const athlete = createAthlete({ name: 'Carol', trailAccess: false })
    expect(athlete.trailAccess).toBe(false)
  })

  it('nullable fields round-trip as null, not undefined', () => {
    const athlete = createAthlete({ name: 'Dave' })
    expect(athlete.email).toBeNull()
    expect(athlete.injuries).toBeNull()
    expect(athlete.recentRaces).toBeNull()
    expect(athlete.coachNotes).toBeNull()
    expect(athlete.raceName).toBeNull()
    expect(athlete.raceDate).toBeNull()
  })
})

describe('getAthlete', () => {
  it('returns null for non-existent id', () => {
    expect(getAthlete(99999)).toBeNull()
  })

  it('returns the athlete for an existing id', () => {
    const created = createAthlete({ name: 'Eve' })
    const fetched = getAthlete(created.id)
    expect(fetched).not.toBeNull()
    expect(fetched!.name).toBe('Eve')
  })
})

describe('listAthletes', () => {
  it('returns empty array when no athletes', () => {
    expect(listAthletes()).toEqual([])
  })

  it('returns all created athletes', () => {
    createAthlete({ name: 'Alice' })
    createAthlete({ name: 'Bob' })
    const athletes = listAthletes()
    expect(athletes).toHaveLength(2)
    expect(athletes.map(a => a.name)).toContain('Alice')
    expect(athletes.map(a => a.name)).toContain('Bob')
  })
})

describe('updateAthlete', () => {
  it('partial update changes only specified fields', () => {
    const athlete = createAthlete({ name: 'Frank', email: 'frank@example.com' })
    const updated = updateAthlete(athlete.id, { name: 'Frank Updated' })
    expect(updated).not.toBeNull()
    expect(updated!.name).toBe('Frank Updated')
    expect(updated!.email).toBe('frank@example.com')
  })

  it('trailAccess false→true persists correctly', () => {
    const athlete = createAthlete({ name: 'Grace', trailAccess: false })
    const updated = updateAthlete(athlete.id, { trailAccess: true })
    expect(updated!.trailAccess).toBe(true)
  })

  it('returns null for missing id', () => {
    expect(updateAthlete(99999, { name: 'Nobody' })).toBeNull()
  })
})

describe('addCoachNote', () => {
  it('sets note directly when coachNotes is null', () => {
    const athlete = createAthlete({ name: 'Henry' })
    expect(athlete.coachNotes).toBeNull()
    const updated = addCoachNote(athlete.id, 'First note')
    expect(updated!.coachNotes).toBe('First note')
  })

  it('appends with newline separator when notes exist', () => {
    const athlete = createAthlete({ name: 'Irene', coachNotes: 'Note 1' })
    const updated = addCoachNote(athlete.id, 'Note 2')
    expect(updated!.coachNotes).toBe('Note 1\nNote 2')
  })

  it('returns null for missing id', () => {
    expect(addCoachNote(99999, 'note')).toBeNull()
  })
})

describe('deleteAthlete', () => {
  it('returns false for non-existent id', () => {
    expect(deleteAthlete(99999)).toBe(false)
  })

  it('returns true and removes row for existing athlete', () => {
    const athlete = createAthlete({ name: 'Jack' })
    expect(deleteAthlete(athlete.id)).toBe(true)
    expect(getAthlete(athlete.id)).toBeNull()
  })
})

describe('toDto boolean conversion', () => {
  it('trail_access=1 → trailAccess=true', () => {
    const athlete = createAthlete({ name: 'Karen', trailAccess: true })
    const fetched = getAthlete(athlete.id)
    expect(fetched!.trailAccess).toBe(true)
  })

  it('trail_access=0 → trailAccess=false', () => {
    const athlete = createAthlete({ name: 'Leo', trailAccess: false })
    const fetched = getAthlete(athlete.id)
    expect(fetched!.trailAccess).toBe(false)
  })
})
